rllib/utils/replay_buffers/multi_agent_episode_replay_buffer.py

```diff
@@ -0,0 +1,671 @@
+    def add(
+        self,
+        episodes: Union[List["MultiAgentEpisode"], "MultiAgentEpisode"],
+    ) -> None:
+
+        episodes: List["MultiAgentEpisode"] = force_list(episodes)
+
+        new_episode_ids: List[str] = []
+        for eps in episodes:
+            new_episode_ids.append(eps.id_)
+            self._num_timesteps += eps.env_steps()
+            self._num_timesteps_added += eps.env_steps()
+
+        # Evict old episodes.
+        eps_evicted: List["MultiAgentEpisode"] = []
+        eps_evicted_ids: List[Union[str, int]] = []
+        eps_evicted_idxs: List[int] = []
+        while (
+            self._num_timesteps > self.capacity
+            and self._num_remaining_episodes(new_episode_ids, eps_evicted_ids) != 1
+        ):
+            # Evict episode.
+            eps_evicted.append(self.episodes.popleft())
+            eps_evicted_ids.append(eps_evicted[-1].id_)
+            eps_evicted_idxs.append(self.episode_id_to_index.pop(eps_evicted_ids[-1]))
+            # If this episode has a new chunk in the new episodes added,
+            # we subtract it again.
+            # TODO (sven, simon): Should we just treat such an episode chunk
+            # as a new episode?
+            if eps_evicted_ids[-1] in new_episode_ids:
+                new_eps_to_evict = episodes[new_episode_ids.index(eps_evicted_ids[-1])]
+                self._num_timesteps -= new_eps_to_evict.env_steps()
+                self._num_timesteps_added -= new_eps_to_evict.env_steps()
+                episodes.remove(new_eps_to_evict)
+            # Remove the timesteps of the evicted episode from the counter.
+            self._num_timesteps -= eps_evicted[-1].env_steps()
+            self._num_agent_timesteps -= eps_evicted[-1].agent_steps()
+            self._num_episodes_evicted += 1
+            # Remove the module timesteps of the evicted episode from the counters.
+            self._evict_module_episodes(eps_evicted[-1])
+
+        # Add agent and module steps.
+        for eps in episodes:
+            self._num_agent_timesteps += eps.agent_steps()
+            self._num_agent_timesteps_added += eps.agent_steps()
+            # Update the module counters by the module timesteps.
+            self._update_module_counters(eps)
+
+        # Remove corresponding indices, if episodes were evicted.
+        if eps_evicted_idxs:
+            new_indices = []
+            for idx_tuple in self._indices:
+                # If episode index is not from an evicted episode, keep it.
+                if idx_tuple[0] not in eps_evicted_idxs:
+                    new_indices.append(idx_tuple)
+            # Assign the new list of indices.
+            self._indices = new_indices
+            # Also remove corresponding module indices.
+            for module_id, module_indices in self._module_indices.items():
+                new_module_indices = []
+                for idx_quad in module_indices:
+                    if idx_quad[0] not in eps_evicted_idxs:
```

Comment:
Can we add a comment explaining, why this [0] index is important to check (and the others aren't)? Explain what the structure of `module_indices` is.